src/microbe_stage/organelle_unlocks/UnlockProgress.cs

```diff
@@ -0,0 +1,126 @@
+    /// </summary>
+    [JsonProperty]
+    private readonly HashSet<OrganelleDefinition> recentlyUnlocked = new();
+
+    [JsonIgnore]
+    private IEnumerable<IUnlockCondition>? allOrganelleUnlockConditions;
+
+    /// <summary>
+    ///   If true, <see cref="IsUnlocked"/> will always return true
+    /// </summary>
+    [JsonProperty]
+    public bool UnlockAll { get; set; }
+
+    [JsonProperty]
+    public GameWorld World { get; set; } = null!;
+
+    public static Type[] GetSupportedStageTypes()
+    {
+        return new[]
+        {
+            typeof(MicrobeStage),
+        };
+    }
+
+    public void InitUnlockConditionTracking()
+    {
+        GetAllUnlockConditions();
+
+        foreach (var constraint in allOrganelleUnlockConditions!)
+            constraint.OnInit(World.StatisticsTracker);
+    }
+
+    /// <summary>
+    ///   Unlock an organelle, returning true if this is the first time it has been unlocked.
+    /// </summary>
+    public bool UnlockOrganelle(OrganelleDefinition organelle, GameProperties game)
+    {
+        if (organelle.UnlockConditions == null || game.FreeBuild || UnlockAll)
+            return false;
+
+        var firstTimeUnlocking = unlockedOrganelles.Add(organelle);
+
+        if (firstTimeUnlocking)
+            recentlyUnlocked.Add(organelle);
+
+        return firstTimeUnlocking;
+    }
+
+    /// <summary>
+    ///   Is the organelle unlocked?
+    /// </summary>
+    public bool IsUnlocked(OrganelleDefinition organelle, GameProperties game, bool autoUnlock)
+    {
+        if (organelle.UnlockConditions == null || game.FreeBuild || UnlockAll)
+            return true;
+
+        if (autoUnlock && organelle.UnlockConditions.Any(unlock => unlock.Satisfied()))
+        {
+            UnlockOrganelle(organelle, game);
+            return true;
+        }
+
+        return unlockedOrganelles.Contains(organelle);
+    }
+
+    /// <summary>
+    ///   Has the organelle been recently unlocked?
+    /// </summary>
+    public bool RecentlyUnlocked(OrganelleDefinition organelle)
+    {
+        return recentlyUnlocked.Contains(organelle);
+    }
+
+    /// <summary>
+    ///   Clear recently unlocked organelles.
+    /// </summary>
+    public void ClearRecentlyUnlocked()
+    {
+        recentlyUnlocked.Clear();
+    }
+
+    public void UpdateUnlockConditionWorldData(ICellProperties? data,
+        EnergyBalanceInfo? energyBalance)
+    {
+        GetAllUnlockConditions();
+
+        foreach (var constraint in allOrganelleUnlockConditions!)
+        {
+            if (constraint is WorldBasedUnlockCondition condition)
+                condition.UpdateWorldData(World, data, energyBalance);
+        }
+    }
+
+    private void GetAllUnlockConditions()
+    {
+        allOrganelleUnlockConditions ??= SimulationParameters.Instance.GetAllOrganelles()
+            .Where(o => o.UnlockConditions != null)
+            .Select(o => o.UnlockConditions)
```

Comment:
Can't this already be the SelectMany line? This seems there's extra line here that doesn't need to be here. Also the same thing applies to `c.Requirements`